src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+
+		@router.post '/v1/purge', (req, res) =>
+			appId = checkString(req.body.appId)
+			force = checkTruthy(req.body.force)
+			if !appId?
+				errMsg = "App not found: an app needs to be installed for purge to work.
+						If you've recently moved this device from another app,
+						please push an app and wait for it to be installed first."
+				return res.status(400).send(errMsg)
+			Promise.using updateLock.lock(appId, { force }), =>
+				@applications.getCurrentApp(appId)
+				.then (app) =>
+					service = app?.services?[0]
+					return res.status(400).send('App not found') if !service?
+					return res.status(400).send('v1 endpoints are only allowed on single-container apps') if app.services.length > 1
+					@applications.executeStepAction(serviceAction('kill', service.serviceId, service, null, skipLock: true), { force })
+					.then =>
+						@applications.executeStepAction({
+							action: 'purge'
+							appId: app.appId
+							options:
+								skipLock: true
+						}, { force })
+					.then =>
+						@applications.executeStepAction(serviceAction('start', service.serviceId, null, service, skipLock: true), { force })
+					.then ->
+						res.status(200).json(Data: 'OK', Error: '')
+			.catch (err) ->
+				res.status(503).send(err?.message or err or 'Unknown error')
+		@router.use(@proxyvisor.router)
+
+module.exports = class ApplicationManager extends EventEmitter
+	constructor: ({ @logger, @config, @db, @eventTracker }) ->
+		@docker = new Docker()
+		@images = new Images({ @docker, @logger, @db })
+		@services = new ServiceManager({ @docker, @logger, @images, @config })
+		@networks = new Networks({ @docker, @logger })
+		@volumes = new Volumes({ @docker, @logger })
+		@proxyvisor = new Proxyvisor({ @config, @logger, @db, @docker, @images, applications: this })
+		@_targetVolatilePerServiceId = {}
+		@validActions = [
+			'kill'
+			'start'
+			'stop'
+			'updateReleaseId'
+			'fetch'
+			'removeImage'
+			'updateImage'
+			'killAll'
+			'purge'
+			'restart'
+			'cleanup'
+			'createNetworkOrVolume'
+			'removeNetworkOrVolume'
+		].concat(@proxyvisor.validActions)
+		@_router = new ApplicationManagerRouter(this)
+		@router = @_router.router
+
+		@images.on('change', @reportCurrentState)
+		@services.on('change', @reportCurrentState)
+
+	serviceAction: serviceAction
+	imageForService: imageForService
+	fetchAction: fetchAction
+
+	reportCurrentState: (data) =>
+		@emit('change', data)
+
+	init: =>
+		@images.cleanupDatabase()
+		.then =>
+			@services.attachToRunning()
+		.then =>
+			@services.listenToEvents()
+
+	# Returns the status of applications and their services
+	getStatus: =>
+		Promise.join(
+			@services.getStatus()
+			@images.getStatus()
+			(services, images) ->
+				apps = {}
+				dependent = {}
+				commit = null
+				# We iterate over the current running services and add them to the current state
+				# of the app they belong to.
+				_.forEach services, (service) ->
```

Comment:
This can be a simple for loop

Reply:
Done (and in a bunch of other places too). I actually like the _.forEach because it abstracts the loop and works consistently for arrays and objects. But I've removed them in almost all places, only left it when iterating over objects or when the use of a function gave a clear benefit (imo) to the resulting code.